Add support for new bootc kickstart command

elkoniu opened the pull request:
Bootc container becomes a new way of OS install supported by the Anaconda. This requires introduction of new kickstart command to support it.

Related discussion: https://github.com/rhinstaller/anaconda/discussions/5197
Resolves: INSTALLER-4024

This is the base for the following change in Anaconda: https://github.com/rhinstaller/anaconda/pull/6298

bcl commented:
I realize this is still a draft :) But a few comments:
* This should be for F43, not F41, unless there are plans to backport this. We generally don't do that in Fedora.
* You don't need placeholder classes for RHEL versions that this won't show up in, like RHEL7 and 8.
* You don't need RHEL classes unless they are different, including commands is controlled by the `pykickstart/handlers/` files.

elkoniu commented:
> I realize this is still a draft :) But a few comments:
> 

Hi Brian, sory for the false alarm - the work is still ongoing :)

>     * This should be for F43, not F41, unless there are plans to backport this. We generally don't do that in Fedora.

Noted and adjusted

> 
>     * You don't need placeholder classes for RHEL versions that this won't show up in, like RHEL7 and 8.
> 
>     * You don't need RHEL classes unless they are different, including commands is controlled by the `pykickstart/handlers/` files.
> 

RHEL related stuff has been removed

> 
> ETA: And there's a typo in the subject line :)

Thanks, fixed

I will continue with draft PR as we are still trying to figure out right set of options for this command. After this will be ready I will ping you and @jkonecny12 for the formal review :)

jkonecny12 commented on pykickstart/commands/bootc.py:
```diff
@@ -0,0 +1,96 @@
+                            """, version=F43, conflicts=self.conflictingCommands)
+        # Rename the osname to stateroot and set default as proposed by
+        # https://github.com/ostreedev/ostree/issues/2794
+        op.add_argument("--stateroot",
+                        version=F43,
+                        help="""
+                        Name for the state directory, also known as "osname".
+                        """)
```
Please set the `default` as the default value (based on the man page).

elkoniu replied:
Done

jkonecny12 commented on pykickstart/commands/bootc.py:
```diff
@@ -0,0 +1,96 @@
+        op.add_argument("--target-imgref",
+                        version=F43,
+                        help="""
+                        Specify the image to fetch for subsequent updates.
```
Please add the note that default is using `--source-imgref`.

elkoniu replied:
Done

jkonecny12 commented on pykickstart/handlers/f43.py:
```diff
@@ -68,6 +68,7 @@ class F43Handler(BaseHandler):
         "timesource": commands.timesource.F33_Timesource,
         "ostreecontainer": commands.ostreecontainer.F38_OSTreeContainer,
         "ostreesetup": commands.ostreesetup.F38_OSTreeSetup,
+        "bootc": commands.bootc.F43_Bootc,
```
The list is sorted alphabetically, please move the `bootc` command.

jkonecny12 commented on pykickstart/commands/__init__.py:
```diff
@@ -64,6 +64,7 @@
     timesource,
     ostreecontainer,
     ostreesetup,
+    bootc,
```
The same here. It should be sorted alphabetically.

jkonecny12 commented:
Also please fix the tests.

jkonecny12 commented:
And please fix the typo `fot` in the commit message :)

jkonecny12 commented:
Hi @cgwalters, could you please verify bootc kickstart command API? We tried to get it close to bootc as possible so it is less confusing to move between the tools. We also removed the `remote` as we were not able to find it in `bootc`, not sure if that is a correct approach.

cgwalters reviewed (approved):
Looks sane to me superficially. Most other options (e.g. `--karg`) are already wrapped by other anaconda verbs.

One thing not covered though is `--bound-images`; but we probably just want to hardcode `--bound-images=pull` for anaconda, it's effectively what the code is doing now anyways.

bcl commented:
I agree with @cgwalters, mostly :) If the command is just being run by anaconda then passing through arguments makes sense. If anaconda needs to take some action, or check the validity of arguments before running it then it makes sense to implement them in kickstart. It's a bit of a balance, but the difference in my mind is whether the command is implemented in anaconda, eg. like the part command.

But also remember that there is some value to implementing arguments, it makes it so that ksvalidator can be used on the kickstart to discover errors before actually using it.

cgwalters commented:
> But also remember that there is some value to implementing arguments, it makes it so that ksvalidator can be used on the kickstart to discover errors before actually using it.

Though bootc is in Rust with clap which is based on a declarative description of the CLI arguments, so in theory we could (like bash completion) generate metadata sufficient to bundle with ksvalidator to do validation there.

(The same applies to a lot of osbuild stage wrappers)

But I dunno, in practice yeah it's nice to be able to validate a kickstart sure, but I think there's so many things that aren't possible to validate that in practice everyone needs to have a good testing framework anyways.

bcl commented:
Yeah, I don't want to go down that road.

jkonecny12 commented on tests/commands/bootc.py:
```diff
@@ -0,0 +1,66 @@
+#
+# Copyright 2023 Red Hat, Inc.
```
Outdated year.

elkoniu replied:
Fixed

jkonecny12 reviewed (approved):
Looks good to me now. Thanks!

jkonecny12 commented:
> Note in the future as bootc evolves I do expect more things to land here, especially [bootc-dev/bootc#22](https://github.com/bootc-dev/bootc/issues/22)
> 
> So one thing I think that would help a lot here is to support a direct passthrough of additional CLI arguments, so we don't need to wrap each individual one in kickstart in the future.
> 
> Prior art for that model looks like https://pykickstart.readthedocs.io/en/latest/kickstart-docs.html#authselect I think?

I like this idea quite a lot, however, I'm afraid how many ways of broken installations we would add this way. For that reason, I would rather implement it on demand honestly. @cgwalters, are you able to tell how fragile the passthrough argument would be?

elkoniu commented:
@bcl I think we have a merge candidate here :)

bcl commented on tests/commands/bootc.py:
```diff
@@ -0,0 +1,66 @@
+
+    def runTest(self):
+        # PASS tests
+        cmdstr = "bootc --source-imgref=quay.io/centos-bootc/centos-bootc:stream9"
```
When writing PASS tests you should also test the output to make sure it is what you expect. In this case the output is different because you are always filling in the target if it isn't set. So it needs to:
* test the output string
* test with setting a different `--target-imgref`

elkoniu replied:
I have added test you required but from some reason it is failing :) I am wondering if this is some string formatting bug.

bcl replied:
Typo in the assert :) `self.assert_parse(cmdstr, cmdstr)` instead of `self.assert_parse(cmdstr, cmdstr_expected)`
Also still needs a test for using different source and target to demonstrate it not getting copied.

elkoniu replied:
Thanks for pointing this out, I have updated the code following you suggestions.

bcl commented on pykickstart/commands/ostreecontainer.py:
```diff
@@ -96,3 +96,8 @@ def parse(self, args):
+class F43_OSTreeContainer(F38_OSTreeContainer):
+    removedKeywords = KickstartCommand.removedKeywords
+    removedAttrs = KickstartCommand.removedAttrs
+    conflictingCommands = ["ostreesetup", "bootc"]
```
Please add bootc to the ostreecontainer conflicting test

elkoniu replied:
Done

bcl commented on pykickstart/commands/ostreesetup.py:
```diff
@@ -100,3 +101,8 @@ class F38_OSTreeSetup(F21_OSTreeSetup):
+class F43_OSTreeSetup(F38_OSTreeSetup):
+    removedKeywords = KickstartCommand.removedKeywords
+    removedAttrs = KickstartCommand.removedAttrs
+    conflictingCommands = ["ostreecontainer", "bootc"]
```
Please add bootc to the ostreesetup conflict test

elkoniu replied:
Done

bcl reviewed (approved):
Looks good. I'm not sure why coveralls is reporting twice, but it doesn't matter.
